File: data/src/main/java/home/oleg/placesnearme/data/mapper/DetailedVenueMapper.java
package home.oleg.placesnearme.data.mapper;

import java.util.ArrayList;
import java.util.List;

import home.oleg.placenearme.models.Contact;
import home.oleg.placenearme.models.DetailedVenue;
import home.oleg.placenearme.models.Photo;
import home.oleg.placesnearme.data.model.DetailedVenueDbEntity;
import home.oleg.placesnearme.data.model.DetailedVenueWithPhotos;

/**
 * Created by Oleg Sheliakin on 25.09.2018.
 * Contact me by email - olegsheliakin@gmail.com
 */
public final class DetailedVenueMapper {

    private DetailedVenueMapper() {
    }

    public static DetailedVenue map(home.oleg.placesnearme.network.models.DetailedVenue detailedVenue) {
        DetailedVenue result = new DetailedVenue();

        if (detailedVenue.getCategories() != null && !detailedVenue.getCategories().isEmpty()) {
            result.setCategory(CategoryMapper.map(detailedVenue.getCategories().get(0)));
        }

        result.setDescription(detailedVenue.getDescription());
        result.setRating(detailedVenue.getRating());
        result.setPhotos(PhotoMapper.map(detailedVenue.getPhotos()));
        result.setLocation(LocationMapper.map(detailedVenue.getLocation()));
        result.setId(detailedVenue.getId());
        result.setName(detailedVenue.getName());
        result.setHours(HoursMapper.map(detailedVenue.getHours()));

        if (detailedVenue.getContact() != null) {
            Contact contact = ContactMapper.map(detailedVenue.getContact());
            result.setContact(contact);
        }

        return result;
    }

    public static DetailedVenueWithPhotos map(DetailedVenue venue) {
        DetailedVenueWithPhotos detailedVenueWithPhotos = new DetailedVenueWithPhotos();

        DetailedVenueDbEntity detailedVenueDbEntity = new DetailedVenueDbEntity();
        detailedVenueDbEntity.setId(venue.getId());
        detailedVenueDbEntity.setTitle(venue.getName());
        detailedVenueDbEntity.setRating(venue.getRating());
        detailedVenueDbEntity.setContact(venue.getContact());

        detailedVenueDbEntity.setCategory(venue.getCategory());
        detailedVenueDbEntity.setLocation(venue.getLocation());
        detailedVenueDbEntity.setDescription(venue.getDescription());
        detailedVenueDbEntity.setFavorite(venue.isFavorite());

        detailedVenueWithPhotos.setVenue(detailedVenueDbEntity);
        detailedVenueWithPhotos.setPhotos(PhotoMapper.mapToDb(venue.getPhotos()));

        return detailedVenueWithPhotos;
    }

    public static List<DetailedVenue> map(List<DetailedVenueWithPhotos> detailedVenueWithPhotos) {
        List<DetailedVenue> detailedVenues = new ArrayList<>();

        for (DetailedVenueWithPhotos detailedVenueWithPhoto : detailedVenueWithPhotos) {
            detailedVenues.add(map(detailedVenueWithPhoto));
        }

        return detailedVenues;
    }

    public static DetailedVenue map(DetailedVenueWithPhotos detailedVenueWithPhotos) {
        DetailedVenueDbEntity venueDbEntity = detailedVenueWithPhotos.getVenue();

        DetailedVenue detailedVenue = new DetailedVenue();
        detailedVenue.setId(venueDbEntity.getId());
        detailedVenue.setName(venueDbEntity.getTitle());
        detailedVenue.setRating(venueDbEntity.getRating());

        detailedVenue.setContact(venueDbEntity.getContact());
        detailedVenue.setLocation(detailedVenue.getLocation());

        detailedVenue.setDescription(venueDbEntity.getDescription());
        detailedVenue.setFavorite(venueDbEntity.isFavorite());

        List<Photo> photoList = PhotoMapper.mapFromDb(detailedVenueWithPhotos.getPhotos());
        detailedVenue.setPhotos(photoList);

        return detailedVenue;
    }
}

File: data/src/main/java/home/oleg/placesnearme/data/mapper/DetailedVenueMapper.kt
package home.oleg.placesnearme.data.mapper

import java.util.ArrayList

import home.oleg.placenearme.models.DetailedVenue
import home.oleg.placesnearme.data.model.DetailedVenueDbEntity
import home.oleg.placesnearme.data.model.DetailedVenueWithPhotos
import home.oleg.placesnearme.data.provider.LocationStore

/**
 * Created by Oleg Sheliakin on 25.09.2018.
 * Contact me by email - olegsheliakin@gmail.com
 */
object DetailedVenueMapper {

    fun map(detailedVenue: home.oleg.placesnearme.network.models.DetailedVenue) = DetailedVenue(
            description = detailedVenue.description,
            name = detailedVenue.name,
            id = detailedVenue.id,
            photos = detailedVenue.photos.map { PhotoMapper.map(it) },
            rating = detailedVenue.rating,
            location = detailedVenue.location.let { LocationMapper.map(it) },
            contact = detailedVenue.contact?.let(ContactMapper::map),
            category = detailedVenue.categories?.firstOrNull()?.let(CategoryMapper::map),
            hours = detailedVenue.hours?.let { HoursMapper.map(it) }
    )

    fun map(venue: DetailedVenue): DetailedVenueWithPhotos {
        val detailedVenue = DetailedVenueDbEntity(
                id = venue.id,
                title = venue.name,
                category = venue.category,
                contact = venue.contact,
                location = venue.location,
                rating = venue.rating,
                isFavorite = venue.isFavorite,
                description = venue.description,
                isHereNow = venue.isHereNow
        )

        val detailedVenueWithPhotos = DetailedVenueWithPhotos(detailedVenue)
        detailedVenueWithPhotos.photos = PhotoMapper.mapToDb(venue.photos)

        return detailedVenueWithPhotos
    }

    fun map(detailedVenueWithPhotos: List<DetailedVenueWithPhotos>): List<DetailedVenue> {
        val detailedVenues = ArrayList<DetailedVenue>()

        for (detailedVenueWithPhoto in detailedVenueWithPhotos) {
            detailedVenues.add(map(detailedVenueWithPhoto))
        }

        return detailedVenues
    }

    fun map(detailedVenueWithPhotos: DetailedVenueWithPhotos): DetailedVenue {
        val venue = detailedVenueWithPhotos.venue

        return DetailedVenue(
                id = venue.id,
                isHereNow = venue.isHereNow,
                description = venue.description,
                isFavorite = venue.isFavorite,
                rating = venue.rating,
                location = venue.location,
                contact = venue.contact,
                category = venue.category,
                photos = PhotoMapper.mapFromDb(detailedVenueWithPhotos.photos),
                name = venue.title
        )
    }
}
